### PR TITLE
Equivalent expressions: assignments [3/n]

### PR DESCRIPTION
**Pull request by kkjeer**
This PR updates the sets UEQ and G of equivalent expressions while checking expressions that involve assignments. This is done with the following changes:

* Update UEQ and G from CheckVarDecl (for declarations with initializers), CheckBinaryOperator (for assignments and compound assignments), and CheckUnaryOperator (for increment/decrement operators).
* Add a GetIncomingBlockState method that gets the state that is true at the beginning of a CFG block by getting the intersection of the UEQ sets of the block's predecessors.
* Add an UpdateAfterAssignment method that updates UEQ and G after an assignment to a variable V.
* Add a GetOriginalValue method that gets the original value of a variable in an expression. Helper methods IsInvertible and Inverse are used to get the original value.
* Add helpers PruneVariableReferences and VariableOccurrenceCount to help with getting the inverse of an expression.
* Add utility classes VariableUtil and ExprCreatorUtil to help with detecting and comparing variables and creating certain kinds of expressions.
* Modify asserts and diagnostic behavior in CanonBounds, TreeTransform, and SemaExpr to reflect the greater variety of expressions that can be lexicographically compared and transformed.

Testing:
* Changed the behavior of the -fdump-checking-state flag to only dump the checking state after checking a top-level CFG statement, rather than after every expression. The equiv-exprs.c test has been modified to reflect this change.
* Added a test equiv-expr-sets.c to test that the set UEQ of sets of equivalent expressions is properly updated while checking assignments. This file also tests that the incoming UEQ set is properly determined for CFG blocks in functions with `if` statements.
* Passed manual testing on Windows.
* Passed automated testing on Windows/Linux.

### CODE REVIEW

File: clang/lib/Sema/SemaBounds.cpp [hunk 1]
```diff
@@ -112,6 +112,90 @@ class BoundsUtil {
 };
 }
 
+namespace {
+  class VariableUtil {
+    public:
+      // If E is a (possibly parenthesized or no-op cast) variable V,
+      // GetVariable returns V.  Otherwise, GetVariable returns nullptr.
+      static DeclRefExpr *GetVariable(Sema &SemaRef, Expr *E) {
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp]: I found this code hard to follow because variable are used in different ways for lvalue and rvalue expressions.  For an assignment, only a possibly-parenthesized DeclRef (an lvalue) occurs on the left-hand side of an operator.  When the value is read, there is an lvalue-to-rvalue cast.

 think you should have versions of GetVariable that are distinct for lvalue vs. rvalue expressions.  For an lvalue expression, the method would check for a possibly-parenthesized DeclRef expressions.  For an rvalue expression, it would check for a Ipossibly-parenthesized lvalue-to-rvalue cast of a possibly-parenthesized DeclRef expression.

IgnoreParenNoopCasts isn't the right method to use here.  LValue-to-rvalue casts actually read memory, and that isn't a no-op cast.

File: clang/lib/Sema/SemaBounds.cpp [hunk 2]
```diff
@@ -112,6 +112,90 @@ class BoundsUtil {
...
+        if (!E)
+          return nullptr;
+        return dyn_cast<DeclRefExpr>(E->IgnoreParenNoopCasts(SemaRef.Context));
+      }
+
+      // SameVariable returns true if the expressions E1 and E2
+      // are the same (possibly parenthesized or no-op cast) variable.
+      static bool SameVariable(Sema &SemaRef, Expr *E1, Expr *E2) {
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp]: If you look at all uses of SameVariable, I believe the first expression is always a DeclRef expression.  I think it would be clearer to change the type of that argument.   You may need specialized versions for lvalue vs. rvalue expressions.

File: clang/lib/Sema/SemaBounds.cpp [hunk 3]
```diff
@@ -2327,8 +2518,50 @@ namespace {
         }
       }
 
+      // Update State.UEQ and State.G.
```
[the remark this answers is not in the document]
Comment (dtarditi) [reply clang/lib/Sema/SemaBounds.cpp:2499]: I didn't see this comment addressed in your most recent change.  It would have been helpful to have a clarifying response here about why it wasn't addressed.

I looked into this and see that UpdateG is screening out non-modifying expressions in the S_i.   I am pretty concerned about this.  Expression equivalence checking assumes that expressions are non-modifying expressions and may produce incorrect results if they are modifying expressions.   

For this reason, I think we should have an invariant that none of the subexpressions G_i containy modifying expression. This change is already large and I'd rather land it now, then have it continune onward.  @kkjeer could you file an issue about this, and follow-up in a subsequent check-in.

File: clang/test/CheckedC/inferred-bounds/equiv-expr-sets.c [hunk 4]
```diff
@@ -0,0 +1,1228 @@
...
+  // CHECK-NEXT: ImplicitCastExpr {{.*}} <LValueToRValue>
+  // CHECK-NEXT:   DeclRefExpr {{.*}} 'i'
+  // CHECK-NEXT: }
+  // CHECK-NEXT: }
+}
+
+// BinaryOperator: compound assignment to a variable
+void f3(unsigned i) {
+  // Updated UEQ: { { (i - 2) + 2, i } }
+  i += 2;
```
Comment (dtarditi) [inline clang/test/CheckedC/inferred-bounds/equiv-expr-sets.c:67]: This first expression in the UEQ set is a tautology.  When updating UEQ information, we should screen out equality facts implied by assignments to variables where the variable also appears on the right-hand side of the assignments.   It's fine to flag this as a follow-up issue.  This is already a large pull request.

File: clang/lib/Sema/SemaBounds.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/lib/Sema/SemaBounds.cpp:2499]: I created issue #815 to track this work.

File: clang/test/CheckedC/inferred-bounds/equiv-expr-sets.c (hunk 4 above)
Comment (kkjeer) [reply clang/test/CheckedC/inferred-bounds/equiv-expr-sets.c:67]: I created issue #814 to track this work.

File: clang/lib/Sema/SemaBounds.cpp [hunk 5]
```diff
@@ -2699,6 +2898,52 @@ namespace {
...
+          // expressions that produce the same value as the variable `e1`,
+          // and these expressions should not be added to UEQ.
+          State.G.clear();
+          Expr *OV = GetOriginalValue(V, RHS, State.UEQ);
+          UpdateAfterAssignment(V, Target, OV, CSS, State, State);
+        }
+
+        // Update the set G of expressions that produce the same value as e.
+        if (RHS) {
+          // For post-inc/dec operators, use the rvalue for `e1` to update G.
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp:2934]: This logic is reversed.   In the state after the expression ++x has executed, the non-modifying expression equivalent to the value produced by ++x is x.   For x++, the non-modifying expression is x - 1..
